### PR TITLE
Feature/kscu 7 cart rebuilding link

### PR DESCRIPTION
**Pull request by dknight-mazegroup**
## Description

This update adds the capability to create a Cart Rebuilding link when the Started Checkout event occurs, and contains a new controller to handle rebuilding the cart with values passed via an encoded URL. 

Important to note _this update does not pass session information to the cart rebuilding link_ - only products, product quantity & selected options for products that were in a customer's cart. Basic error handling has been included as part of this update. 

Jira Ticket:

[KSCU-7:](https://themazegroup.atlassian.net/browse/KSCU-7) Add Cart Rebuilding link to Started Checkout event.

## Manual Testing Steps

This has been manually tested in across SFRA & SiteGen in various browsers to confirm consistency. Multiple products have been tested during development, including:

- Standard Products without special options _(ex: clothing, bags, etc.)_
- Standard Products with special options _(ex: Televisions, GPS, etc.)_
- Product Bundles w & w/o special options selected _(ex: gaming consoles, etc.)_
- Full Product Sets _(ex: winter outfits, spring outfits, etc.)_

Some test links with pre-populated products been have been added below - these are best tested in fresh incognito windows:

- [SiteGen: 6 products](https://zzqk-005.dx.commercecloud.salesforce.com/on/demandware.store/Sites-SiteGenesis-Site/default/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoiNzQwMzU3MzU4MDk1IiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6IjcwMTY0Mjg2NzEwNCIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI3MDE2NDI4NTM2OTUiLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W119LHsicHJvZHVjdElEIjoibWljcm9zb2Z0LXhib3gzNjAtYnVuZGxlIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoiY29uc29sZVdhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMiIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiAzIFllYXIgV2FycmFudHkifV19LHsicHJvZHVjdElEIjoiYXBwbGUtaXBvZC10b3VjaDMyZ2IiLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W119LHsicHJvZHVjdElEIjoiNzQ5NzQzMTAtMSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX1d) - prod. set, prod. w/ selected option, prod. w/ selected option, prod. w/o  option
- [SiteGen: 2 products](https://zzqk-005.dx.commercecloud.salesforce.com/on/demandware.store/Sites-SiteGenesis-Site/default/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoiMDI5NDA3MzMxMjI3IiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6Im1pY3Jvc29mdC14Ym94MzYwLWJ1bmRsZSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbeyJvcHRpb25JRCI6ImNvbnNvbGVXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDAiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogTm9uZSJ9XX1d) - one product bundle w/o extra option and one product w/o option
- [SFRA: 14 products](https://zzys-003.dx.commercecloud.salesforce.com/on/demandware.store/Sites-RefArch-Site/default/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoic29ueS1wczMtYnVuZGxlTSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbeyJvcHRpb25JRCI6ImNvbnNvbGVXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDAiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogTm9uZSJ9XX0seyJwcm9kdWN0SUQiOiJ2aXppby12bzM3bE0iLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W3sib3B0aW9uSUQiOiJ0dldhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMyIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiA1IFllYXIgV2FycmFudHkifV19LHsicHJvZHVjdElEIjoiNzAxNjQ0MDU5MjQ4TSIsInF1YW50aXR5IjoyLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI3MDE2NDI4MTk0NDhNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6IjAxMzc0MjMzNTQyMk0iLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W119LHsicHJvZHVjdElEIjoiNzQwMzU3Mzc3MDM0TSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI2OTMwOTI4NE0tMSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiI4ODMzNjA1NDQyMTJNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfSx7InByb2R1Y3RJRCI6IlAwMTM4TSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbXX0seyJwcm9kdWN0SUQiOiJtaWNyb3NvZnQteGJveDM2MC1idW5kbGVNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoiY29uc29sZVdhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMCIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiBOb25lIn1dfSx7InByb2R1Y3RJRCI6Im1pdHN1YmlzaGktbHQtNDAxNDhNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoidHZXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDMiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogNSBZZWFyIFdhcnJhbnR5In1dfSx7InByb2R1Y3RJRCI6Imdhcm1pbi1udXZpLTUwMDBNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOlt7Im9wdGlvbklEIjoiZ3BzV2FycmFudHkiLCJvcHRpb25WYWx1ZUlEIjoiMDAyIiwibGluZUl0ZW1UZXh0IjoiRXh0ZW5kZWQgV2FycmFudHk6IDMgWWVhciBXYXJyYW50eSJ9XX0seyJwcm9kdWN0SUQiOiI4ODMzNjA1MjAzMDhNIiwicXVhbnRpdHkiOjEsIm9wdGlvbnMiOltdfV0=) - 2 prod. sets, 2 prod. bundles, 3 prods. w/ selected option, 1 prod. w/o option
- [SFRA: 2 products](https://zzys-003.dx.commercecloud.salesforce.com/on/demandware.store/Sites-RefArch-Site/en_US/KlaviyoRecreate-Cart?items=W3sicHJvZHVjdElEIjoic29ueS1wczMtYnVuZGxlTSIsInF1YW50aXR5IjoxLCJvcHRpb25zIjpbeyJvcHRpb25JRCI6ImNvbnNvbGVXYXJyYW50eSIsIm9wdGlvblZhbHVlSUQiOiIwMDAiLCJsaW5lSXRlbVRleHQiOiJFeHRlbmRlZCBXYXJyYW50eTogTm9uZSJ9XX0seyJwcm9kdWN0SUQiOiJ2aXppby12bzM3bE0iLCJxdWFudGl0eSI6MSwib3B0aW9ucyI6W3sib3B0aW9uSUQiOiJ0dldhcnJhbnR5Iiwib3B0aW9uVmFsdWVJRCI6IjAwMyIsImxpbmVJdGVtVGV4dCI6IkV4dGVuZGVkIFdhcnJhbnR5OiA1IFllYXIgV2FycmFudHkifV19XQ==) - one product bundle w/o extra option and one one product w/ selected option

## OOTB SFRA & SiteGen Estimated Total Differences

During development, it was noted the estimated totals displayed in the Cart vary slightly between SFRA & SiteGen. SFRA will display a default shipping, sales tax and estimated total as well as the price of individual line items in the cart. SiteGen, however, does not include Shipping and Sales Tax when viewing the cart for the first time or when clicking checkout for the first time _(shipping and taxes are added after proceeding as a guest / logged in user and proceeding through the steps)._ 

The links to rebuild the cart replicate this and display the prices similarly - the SFRA cart includes estimated total with shipping & taxes and individual line item costs whereas the site gen site includes individual line item costs and an estimated total without the shipping and taxes _(as though the cart is viewed for the first time)_. Noting because this appears to be the expected difference between the sites. 

## Pre-Submission Checklist:

[ x ] This update successfully builds
[ x ] No console errors are displayed on the screen with these updates
[ x ] No server-side errors / DW with ease issues are displayed with these updates
[ x ] Console Logs created during development for debugging that are no longer needed have been removed
[ n/a ] ToDo comments were left as necessary for future reference
[ x ] A description of this update & references to the manual tests has been included in this PR

### CODE REVIEW

File: cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        var test = error;
+        res.renderJSON({
+            success: false,
+            error: true,
+            errorMessage: `ERROR - ${error.message}. Please check encoded URL Object and the Cart Model reference.`
+        });
+        return;
+    }
+
+    if (cart && items.length) {
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js:36]: I would recommend adding a check for items before checking items.length since it appears that null is a possible value for items

File: cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            app.getView('Cart', {
+                cart: cart,
+                BonusDiscountLineItem: renderInfo.BonusDiscountLineItem
+            }).render(renderInfo.template);
+        } else {
+            response.redirect(URLUtils.url('Cart-Show'));
+        }
+    }
+
+    if (!cart) {
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js:64]: Add a check for !items as well since it could be null (and items.length == 0)

Comment (kvanandel-maze) [reply cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js:64]: Alternately, checking !cart || !items || !items.length before line 36, return straightaway (with error JSON) and skip the check currently at line 36.

File: cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        res.renderJSON({
+            success: false,
+            error: true,
+            errorMessage: `ERROR - ${error.message}. Please check encoded URL Object and the Cart Model reference.`
+        });
+        return;
+    }
+
+    if (cart && items.length) {
+        var renderInfo = recreateHelpers.addProductToCart(items, cart)
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js:37]: Probably not causing an error, but for completeness add the missing ";" at the end of this line (var renderInfo = ....)

File: cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js [hunk 4]
```diff
@@ -12,6 +13,7 @@ function getData(currentBasket) {
 
     var data = {};
     var basketItems = currentBasket.getProductLineItems().toArray();
+    var reconstructCartItems = [];
     // Create some top-level event data
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js:17]: Are these comments intended to guide the integrator for customizations? If not, can they be removed?

File: cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js [hunk 5]
```diff
@@ -22,11 +24,20 @@ function getData(currentBasket) {
     data.Categories = [];
     data.Items = [];
     data.$email = currentBasket.customerEmail;
+    data.cartRebuildingLink = URLUtils.abs('KlaviyoRecreate-Cart').toString() + `?items=${reconstructCartItems}`;
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js:27]: It looks like the value of reconstructCartItems here is an empty array. Should this be stringified and/or encoded before appending to the URL?

Comment (kvanandel-maze) [reply cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js:27]: nvm.... just noticed the backticks

File: cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js [hunk 6]
```diff
@@ -35,9 +46,12 @@ function getData(currentBasket) {
             data.line_items.push(productObj);
             data.Categories.push.apply(data.Categories, data.line_items[itemIndex].Categories);
             data.Items.push(data.line_items[itemIndex]['Product Name']);
+
+            reconstructCartItems.push({ productID: currentProductID, quantity, options });
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js:50]: Should this object be {productID: currentProductID, quantity: quantity, options: options}? I just had some issues with an object in compatibility mode 21.7 where I had module.exports = {function1, function2}; and had to update it to module.exports = {function1: function1, function2: function2}; It had worked fine in 22.7 mode.

File: cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            errorMessage: `ERROR - Current Basket is: ${currentBasket}. Please check the Current Basket and try again.`,
+            redirectUrl: URLUtils.url('Cart-Show').toString()
+        });
+        return next();
+    };
+
+    // Clean the basket to prevent product duplication on page refresh
+    if (currentBasket && currentBasket.productQuantityTotal > 0) {
+        for (let i = 0; i < currentBasket.productLineItems.length; i++) {
+            let lengthAtIteration = currentBasket.productLineItems.length;
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js:58]: The logic has me a bit confused. I think you could accomplish the goal here by simply stepping backwards through the cart instead using (let i = currentBasket.productLineItems.length-1; i >= 0; i--) or similar? Incrementing and decrementing i makes this a bit hard to follow and debug.

Comment (kvanandel-maze) [reply cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js:58]: I'm also wondering why this logic was not needed in recreateCartHelpers.js addProductToCart. If cart.object addProductToCart is the raw SFCC LineItemCtnr API object, could you create an emptyCart(lineItemCtnr) helper function that could be used by both SFRA and sitegen/controllers? Seems like the LineItemCtnr.allProductLineItems used in addProductToCart could also be used here in lieu of LineItemCtnr.productLineItems to avoid the PromotionMgr call at line 69. Not sure why the shipment removal logic is needed here, but not in addProductToCart, unless that's already included in the sitegen removeProductLineItem logic (not viewable in this PR).

I'm looking at modularity and shared code as a way to potentially make maintenance and bug fixes easier down the road. If it doesn't work to share the helper function, it is what it is.

File: cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            for (let i = 0; i < items.length; i++) {
+                var productToAdd = ProductMgr.getProduct(items[i].productID);
+                var childProducts = productToAdd.bundledProducts ? collections.map(productToAdd.bundledProducts, function (product) { return { pid: product.ID, quantity: null } }) : [];
+                var options = [];
+                items[i].options.forEach(optionObj => {
+                    options.push({ lineItemText: optionObj.lineItemText, optionId: optionObj.optionID, selectedValueId: optionObj.optionValueID});
+                })
+
+                var shipments = Array.from(currentBasket.shipments);
+                shippingHelper.ensureShipmentHasMethod(shipments[0]);
```
Comment (kvanandel-maze) [inline cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js:84]: Is there a reason why ALL shipments in the array shouldn't be checked for a shipping method? Or is there an assumption here that there's only 1 shipment? If the latter, why the array? Just grab defaultShipment and make sure it has a shipping method.

File: cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js (hunk 6 above)
Comment (dknight-mazegroup) [reply cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js:50]: Thanks for this and for showing me how the issue occurred during our call. Just for tracking, reference - confirming this change was made as part of the [follow-up PR](https://github.com/maze-consulting/SFCC_Klaviyo_Maze/pull/17/files). 

That line now reads as: 

`reconstructCartItems.push({ productID: currentProductID, quantity: quantity, options: options });`

File: cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js (hunk 4 above)
Comment (dknight-mazegroup) [reply cartridges/int_klaviyo_core/cartridge/scripts/klaviyo/eventData/startedCheckout.js:17]: was just left in during earlier development on this project. will remove.

File: cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js (hunk 2 above)
Comment (dknight-mazegroup) [reply cartridges/int_klaviyo/cartridge/controllers/KlaviyoRecreate.js:64]: Thanks for this - just added this adjustment and adjustments from your following comments into a new PR that's available here:

- https://github.com/maze-consulting/SFCC_Klaviyo_Maze/pull/19

File: cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js (hunk 7 above)
Comment (dknight-mazegroup) [reply cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js:58]: Thanks - have reworked following our chat on 04.24. Changes related to this are now available in the following PR:

- https://github.com/maze-consulting/SFCC_Klaviyo_Maze/pull/19

File: cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js (hunk 8 above)
Comment (dknight-mazegroup) [reply cartridges/int_klaviyo_sfra/cartridge/controllers/KlaviyoRecreate.js:84]: No longer creating an array - instead iterating through object directly to remain more dynamic and calling the shippingHelper function on every shipment that already exists. That adjustment is available in an updated PR here:

- https://github.com/maze-consulting/SFCC_Klaviyo_Maze/pull/19
